src/arithmetic/big_gmp.rs

```diff
@@ -0,0 +1,131 @@
+/*
+    Multi-party ECSDA
+
+    Copyright 2018 by Kzen Networks
+
+    This file is part of Multi-party ECDSA library
+    (https://github.com/KZen-networks/multi-party-ecdsa)
+
+    Multi-party ECSDA is free software: you can redistribute
+    it and/or modify it under the terms of the GNU General Public
+    License as published by the Free Software Foundation, either
+    version 3 of the License, or (at your option) any later version.
+
+    @license GPL-3.0+ <https://github.com/KZen-networks/multi-party-ecdsa/blob/master/LICENSE>
+*/
+
+use super::gmp::mpz::Mpz;
+use super::traits::{ Converter, Modulo, Samplable };
+use super::rand::{OsRng, Rng};
+
+use std::borrow::Borrow;
+
+pub type BigInteger = Mpz;
+
+impl Converter for Mpz {
+    fn to_vec(value: &Mpz) -> Vec<u8> {
+        let bytes: Vec<u8> = value.borrow().into();
+        bytes
+    }
+}
+
+// TODO: write unit test
+impl Modulo for Mpz {
+    fn mod_pow(base: &Self, exponent: &Self, modulus: &Self) -> Self {
+        base.powm(exponent, modulus)
+    }
+
+    fn mod_mul(a: &Self, b: &Self, modulus: &Self) -> Self {
+        (a.mod_floor(modulus) * b.mod_floor(modulus)).mod_floor(modulus)
```

Comment:
added to: https://github.com/KZen-networks/multi-party-ecdsa/issues/6

Reply:
@mortendahl if the impact of having reductions is negligible why not keep it? 
This function implements modulo arithmetic multiplication rule: https://www.khanacademy.org/computing/computer-science/cryptography/modarithmetic/a/modular-multiplication

Reply:
I tried to chase it down for GMP and Ramp, and they both seem to have some sort of test in place for short circuiting (eg https://github.com/Aatch/ramp/blob/master/src/ll/div.rs#L260-L363) in case the division is useless, but both also do some processing before that could be avoided.

Quick bench using large numbers showed 50% slow-down but this will probably vary a lot based on the size and nature of numbers that are being computed on.

```
test tests::bench_doublemod      ... bench:         764 ns/iter (+/- 86)
test tests::bench_doublemodcheck ... bench:         615 ns/iter (+/- 24)
test tests::bench_singlemod      ... bench:         449 ns/iter (+/- 38)
```

Rust code:
```rust
#![feature(test)]

extern crate test;
extern crate gmp;

#[cfg(test)]
mod tests {

    use test::Bencher;

    use std::str::FromStr;
    use gmp::mpz::Mpz;

    #[bench]
    fn bench_singlemod(b: &mut Bencher) {

        let n = Mpz::from_str("23601375460155562757123678360900229644381030159964965932095920363097284825175029196457022864038449469086188985762066259059164844287276915193108505099612427967057134520230945630209577834878763915645946525724125804370016991193585261991964913084246563304755455418791629494251095184144084978275430600444710605147457044597210354635288909909182640243950968376955162386281524128586829759108414295175173359174297599533960370415928328418610692822180389889327103292184546896322100484378149887147731744901289563127581082141485046742100147976163228583170704180024449958168221243717383276594270459874555884125566472776234343167371").unwrap();
        let p = Mpz::from_str("34832330983979593310074301486537017973460461278300587514468301043894574906886127642530475786889672304776052879927627556769456140664043088700743909632312483413393134504352834240399191134336344285483935856491230340093391784574980688823380828143810804684752914935441384845195613674104960646037368551517").unwrap();
        let q = Mpz::from_str("7007245654463598139927898730476924736461654463975966787719309357536545869203069369466212089132653564188443272208127277664424448947476335413293018778018615899291704693105620242763173357203898195318179150836424196645745308205164116144020613415407736216097185962171301808761138424668335445923774195463").unwrap();

        b.iter(|| {
            let _ = (&p * &q) % &n;
        })
    }

    #[bench]
    fn bench_doublemod(b: &mut Bencher) {

        let n = Mpz::from_str("23601375460155562757123678360900229644381030159964965932095920363097284825175029196457022864038449469086188985762066259059164844287276915193108505099612427967057134520230945630209577834878763915645946525724125804370016991193585261991964913084246563304755455418791629494251095184144084978275430600444710605147457044597210354635288909909182640243950968376955162386281524128586829759108414295175173359174297599533960370415928328418610692822180389889327103292184546896322100484378149887147731744901289563127581082141485046742100147976163228583170704180024449958168221243717383276594270459874555884125566472776234343167371").unwrap();
        let p = Mpz::from_str("34832330983979593310074301486537017973460461278300587514468301043894574906886127642530475786889672304776052879927627556769456140664043088700743909632312483413393134504352834240399191134336344285483935856491230340093391784574980688823380828143810804684752914935441384845195613674104960646037368551517").unwrap();
        let q = Mpz::from_str("7007245654463598139927898730476924736461654463975966787719309357536545869203069369466212089132653564188443272208127277664424448947476335413293018778018615899291704693105620242763173357203898195318179150836424196645745308205164116144020613415407736216097185962171301808761138424668335445923774195463").unwrap();

        b.iter(|| {
            let _ = ((&p % &n) * (&q % &n)) % &n;
        })
    }

    #[bench]
    fn bench_doublemodcheck(b: &mut Bencher) {

        let n = Mpz::from_str("23601375460155562757123678360900229644381030159964965932095920363097284825175029196457022864038449469086188985762066259059164844287276915193108505099612427967057134520230945630209577834878763915645946525724125804370016991193585261991964913084246563304755455418791629494251095184144084978275430600444710605147457044597210354635288909909182640243950968376955162386281524128586829759108414295175173359174297599533960370415928328418610692822180389889327103292184546896322100484378149887147731744901289563127581082141485046742100147976163228583170704180024449958168221243717383276594270459874555884125566472776234343167371").unwrap();
        let p = Mpz::from_str("34832330983979593310074301486537017973460461278300587514468301043894574906886127642530475786889672304776052879927627556769456140664043088700743909632312483413393134504352834240399191134336344285483935856491230340093391784574980688823380828143810804684752914935441384845195613674104960646037368551517").unwrap();
        let q = Mpz::from_str("7007245654463598139927898730476924736461654463975966787719309357536545869203069369466212089132653564188443272208127277664424448947476335413293018778018615899291704693105620242763173357203898195318179150836424196645745308205164116144020613415407736216097185962171301808761138424668335445923774195463").unwrap();

        b.iter(|| {
            let p_norm: Mpz = if &p < &n { p.clone() } else { &p % &n };
            let q_norm: Mpz = if &q < &n { q.clone() } else { &q % &n };            
            let _ = (p_norm * q_norm) % &n;
        })
    }

}
```